File: OEM/Domain/OperatorPlans/AlgorithmUsed.cs
namespace OEMAPI.Domain.OperatorPlans
{
    public class AlgorithmUsed : ValueObject
    {
        public string Value { get; }

        public AlgorithmUsed(string value)
        {
            if (string.IsNullOrWhiteSpace(value))
                throw new BusinessRuleValidationException("Algorithm Used cannot be empty.");

            Value = value;
        }

        protected override IEnumerable<object> GetEqualityComponents()
        {
            yield return Value;
        }
    }
}

File: OEM/src/Domain/OperatorPlans/AlgorithmUsed.ts
import { ValueObject } from "../core/domain/ValueObject";
import { Result } from "../core/logic/Result";
import { Guard } from "../core/logic/Guard";

interface AlgorithmUsedProps {
  value: string;
}

export class AlgorithmUsed extends ValueObject<AlgorithmUsedProps> {

  get value(): string {
    return this.props.value;
  }

  private constructor(props: AlgorithmUsedProps) {
    super(props);
  }

  public static create(algorithm: string): Result<AlgorithmUsed> {
    const guardResult = Guard.againstNullOrUndefined(algorithm, "algorithmUsed");

    if (!guardResult.succeeded) {
      return Result.fail<AlgorithmUsed>(guardResult.message);
    }

    return Result.ok<AlgorithmUsed>(
      new AlgorithmUsed({ value: algorithm })
    );
  }
}
